SwiftGit2/Repository.swift

```diff
@@ -558,4 +582,234 @@ final public class Repository {
+																							 self.pointer,
+																							 nil,
+																							 tree,
+																							 nil)
+				return processTreeToTreeDiff(diffResult, diff: diff)
+			})
+		}
+
+		return .failure(NSError(gitError: -1, pointOfFailure: "diff(from: to:)"))
+	}
+
+	private func processTreeToTreeDiff(_ diffResult: Int32, diff: OpaquePointer?) -> Result<Diff, NSError> {
+		guard diffResult == GIT_OK.rawValue else {
+			return .failure(NSError(gitError: diffResult,
+															pointOfFailure: "git_diff_tree_to_tree"))
+		}
+
+		let diffObj = Diff(diff!)
+		git_diff_free(diff)
+		return .success(diffObj)
+	}
+
+	private func processDiffDeltas(_ diffResult: OpaquePointer) -> Result<[Diff.Delta], NSError> {
+		typealias Delta = Diff.Delta
+		var returnDict = [Delta]()
+
+		let count = git_diff_num_deltas(diffResult)
+
+		for i in 0..<count {
+			let delta = git_diff_get_delta(diffResult, i)
+			let gitDiffDelta = Diff.Delta((delta?.pointee)!)
+
+			returnDict.append(gitDiffDelta)
+		}
+
+		let result = Result<[Diff.Delta], NSError>.success(returnDict)
+		return result
+	}
+
+	private func safeTreeForCommitId(_ oid: OID) -> Result<Tree, NSError> {
+		return withGitObject(oid, type: GIT_OBJ_COMMIT) { commit in
+			let treeId = git_commit_tree_id(commit)
+			let tree = self.tree(OID(treeId!.pointee))
+			guard tree.error == nil else {
+				return .failure(tree.error!)
+			}
+			return tree
+		}
+	}
+
+	/// Caller responsible to free returned tree with git_object_free
+	private func unsafeTreeForCommitId(_ oid: OID) -> Result<OpaquePointer, NSError> {
+		var commit: OpaquePointer? = nil
+		var oid = oid.oid
+		let commitResult = git_object_lookup(&commit, self.pointer, &oid, GIT_OBJ_COMMIT)
+		guard commitResult == GIT_OK.rawValue else {
+			return .failure(NSError(gitError: commitResult, pointOfFailure: "git_object_lookup"))
+		}
+
+		var tree: OpaquePointer? = nil
+		let treeId = git_commit_tree_id(commit)
+		let treeResult = git_object_lookup(&tree, self.pointer, treeId, GIT_OBJ_TREE)
+
+		git_object_free(commit)
+
+		guard treeResult == GIT_OK.rawValue else {
+			return .failure(NSError(gitError: treeResult, pointOfFailure: "git_object_lookup"))
+		}
+
+		return Result<OpaquePointer, NSError>.success(tree!)
+	}
+
+	// MARK: - Status
+
+	public func status() -> Result<[StatusEntry], NSError> {
+		var returnArray = [StatusEntry]()
+
+		// Do this because GIT_STATUS_OPTIONS_INIT is unavailable in swift
+		let pointer = UnsafeMutablePointer<git_status_options>.allocate(capacity: 1)
+		let optionsResult = git_status_init_options(pointer, UInt32(GIT_STATUS_OPTIONS_VERSION))
+		guard optionsResult == GIT_OK.rawValue else {
+			return .failure(NSError(gitError: optionsResult, pointOfFailure: "git_status_init_options"))
+		}
+		var options = pointer.move()
+		pointer.deallocate(capacity: 1)
+
+		var unsafeStatus: OpaquePointer? = nil
+		let statusResult = git_status_list_new(&unsafeStatus, self.pointer, &options)
```

Comment:
One last memory issue: This needs a `git_status_list_free` in a `defer`.

Reply:
👍